### PR TITLE
TVB-2428: Handle UI Filters

### PR DESCRIPTION
**Pull request by VinczeRobert**
I managed to return only the options from the Python side and replace them in the JS side. However, there is a problem that I still need to solve, and that is the fact that default None or All Options are deleted this way, so I guess I can't just delete all options and put the new ones, I have to keep those default options. Also probably some parts of the code could look better, but I just wanted to make the PR now so I can get a review if the way this went is a good one or not.

### CODE REVIEW

File: framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py [hunk 1]
```diff
@@ -305,6 +306,8 @@ def index(self):
                                                           last_request_type=cherrypy.request.method,
                                                           is_first_fragment=True)
 
+        common.add2session(KEY_ADAPTER, '13')
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py:309]: This should not be ...

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py [hunk 2]
```diff
@@ -377,64 +383,40 @@ def accepted__aggregation_functions():
...
+        filter = FilterChain(fields=fields, operations=operations, values=values)
+        project = common.get_current_project()
+
+        form = Form(project_id=project.id, draw_ranges=True)
+        data_type_gid_attr = DataTypeGidAttr(linked_datatype=REGISTRY.get_datatype_for_index(index_class()))
+        select_field = TraitDataTypeSelectField(data_type_gid_attr, form, conditions=filter)
+
+        options_response = ''
+        for option in select_field.options():
+            options_response = options_response + '<option value=' + '"' + option.value + '"' + ' class="form-control">' + option.label + '</option>'
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:417]: This HTML string concatenation is not good to have.
Can't you return a json which is parsed in the JS and converted in HTML code there on the client ?
Or even better use a Jinja template to render the response as HTML

File: framework_tvb/tvb/interfaces/web/static/js/filters.js [hunk 3]
```diff
@@ -124,44 +123,15 @@ function refreshData(parentDivId, divSufix, name, sessionStoredTreeKey, gathered
...
-            } else {
-                displayMessage("Filter could not be applied!" + name, "infoMessage");
-                return;
-            }
-            displayMessage("Filters applied...");
+        type: 'POST',
+        url: "/flow/get_filtered_datatypes/" + name + '/' + datatypeIndex + '/' + $.toJSON(gatheredData),
+        success: function (response) {
+            var select_field = document.getElementById(name);
+            select_field.innerHTML = response;
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/static/js/filters.js:132]: Basically here compose the options (from JSON response), if they can not be returned by a Jinja template

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py [hunk 4]
```diff
@@ -241,6 +245,8 @@ def default(self, step_key, adapter_key, cancel=False, back_page=None, not_reset
             template_specification[common.KEY_DISPLAY_MENU] = True
             template_specification[common.KEY_BACK_PAGE] = back_page_link
 
+        common.add2session(common.KEY_ADAPTER, adapter_key)
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:248]: Do we still need this in session ?

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py [hunk 5]
```diff
@@ -377,64 +383,40 @@ def accepted__aggregation_functions():
...
+            fields.append(filters_dict['fields'][idx])
+            operations.append(filters_dict['operations'][idx])
+            values.append(filters_dict['values'][idx])
+
+        filter = FilterChain(fields=fields, operations=operations, values=values)
+        project = common.get_current_project()
+
+        form = Form(project_id=project.id, draw_ranges=True)
+        data_type_gid_attr = DataTypeGidAttr(linked_datatype=REGISTRY.get_datatype_for_index(index_class()))
+        select_field = TraitDataTypeSelectField(data_type_gid_attr, form, conditions=filter)
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:413]: The new logic looks much cleaner that the prev code! Good job!

Not sure if it is really justified here, but could we potentially push some of this logic in a service and leave here in the controller only the part about inputs parsing and link with a jinja/response json part?

**Comment by liadomide** [comment]
> I managed to return only the options from the Python side and replace them in the JS side. However, there is a problem that I still need to solve, and that is the fact that default None or All Options are deleted this way, so I guess I can't just delete all options and put the new ones, I have to keep those default options. Also probably some parts of the code could look better, but I just wanted to make the PR now so I can get a review if the way this went is a good one or not.

None option would be simple.
ALL should also be replaced after filtering, as it will contain a new set (the GUID after filtering) instead of the ones before filtering.

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html [hunk 6]
```diff
@@ -1,14 +1,28 @@
-<p class="field-data">
+<div class="field-data">
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:1]: why this change from `p` into `div` is necessary ?
it will break consistency with the other fields...

File: framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py (hunk 1 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py:309]: Yeah, I intented to eliminate that, thanks for the reminder.

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 2 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:417]: JSON would work while using some template, so should I create a new Jinja template for this?

File: framework_tvb/tvb/interfaces/web/static/js/filters.js (hunk 3 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/static/js/filters.js:132]: I think I can't have a JSON response without using some template, when I tried to return for example {'options': string_of_options} only the string options appeared in the response.

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 4 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:248]: No

File: framework_tvb/tvb/interfaces/web/static/js/filters.js (hunk 3 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/static/js/filters.js:132]: in this case (when the return value is a json) see to have the correct annotation(s) on the controller method also...
there are multiple examples in the code already and working

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 5 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:413]: I think we could, why not... But in which one? OperationService? Or a new service just for filters?

Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:413]: flowservice ?

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html (hunk 6 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:1]: Well, when I did that the div added by me didn't want to appear where I wanted and I read on the internet that block elements such as div  should be inside of inline elements such as paragraph (not that it can't be done, it's just not recommented) and changing it to div helped my problem, but now I tried again with <p> and for some reason now it works, so I guess we can leave it as a paragraph.

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 2 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:417]: yes, either a new jinja template, and then you use @using_template or return a dict and then use annotation @jsonify on the controller method

File: framework_tvb/tvb/interfaces/web/static/js/filters.js (hunk 3 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/static/js/filters.js:132]: if you choose annotation `@ using_template` in the controller then you expect a HTML string here (as you do right now) to put directly as innerHTML. If you choose  `@ jsonify` in the controller, then in the JS would be something like:
` r = $.parseJSON(r);
    if (r['status'] === 'ok') ...
`

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 5 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:413]: I don't think we have a Flow Service, so should I create it?

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 2 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:417]: This part is solved.

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 5 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:413]: Wasn't it recently removed?

**Comment by VinczeRobert** [comment]
I believe the general flow is okay now. Tomorrow I really hope I can solve the particular issues as well, like for example filters for simulation_name on TimeSeries don't work as they should.

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 5 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:282]: `FlowService` was recently renamed into `AlgorithmService` ..... by me ;-)

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 4 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:248]: Then, I suggest you remove this line

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py [hunk 7]
```diff
@@ -241,69 +245,43 @@ def default(self, step_key, adapter_key, cancel=False, back_page=None, not_reset
...
         self.fill_default_attributes(template_specification, algorithm.displayname)
         return template_specification
 
-    # @expose_fragment("flow/type2component/datatype2select_simple")
-    def getfiltereddatatypes(self, name, parent_div, tree_session_key, filters):
-        # TODO: fix this use-case
+    @expose_fragment('form_fields/options_field')
+    @settings
+    @context_selected
+    def get_filtered_datatypes(self, datatype_index_path, filters):
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:258]: Wouldn't be simpler the code if you would pass the `dt_module` and `dt_class` strings here?

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 5 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:282]: If I read this again, most of the logic is parsing of strings into filters, so it would be right to stay in Controller.

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html [hunk 8]
```diff
@@ -1,13 +1,30 @@
-<p class="field-data">
-<select name="{{ field.name }}" id="{{ field.name }}" class="dataset-selector" tabindex="1">
+<script type="text/javascript" src="/static/js/filters.js?{{ currentVersionJS }}"></script>
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:1]: Can't you put the import of `filters.js` on a higher level, e.g. in other jinja template where `datatype_select_field.html` gets included also?
If you put it here, it will be included multiple times, as we might have the fragment `datatype_select_field.html` repeating on the same page.

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html [hunk 9]
```diff
@@ -1,13 +1,30 @@
...
 {% for option in field.options() %}
     <option value="{{ option.value }}" {{ 'selected' if option.checked }} class="form-control">{{ option.label }}</option>
 {% endfor %}
 </select>
 {% if field.draw_dynamic_conditions_buttons %}
-    <p class="field-adapters" id="{{ field.owner.name }}{{ field.name }}data_select">
-        <input type="button" tabindex='1' value="Add Filter" onclick="addFilter('{{ field.owner.name }}' + '{{ field.name }}' + 'data_select', {{ field.get_dynamic_filters }})"/>
-        <input type="button" tabindex='1' value="Apply Filters" onclick="refreshData('{{ field.owner.name }}', '{{ field.name }}' + 'data_select', '{{ field.name }}', 'defaultTree')"/>
-    </p>
+    <div class="field-adapters" id="{{ field.owner.name }}{{ field.name }}data_select">
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:11]: Here again, I would prefer if you do not change the tag from `<p>` into `<div>` unless there is something really not working, otherwise the current CSS schema might break

File: framework_tvb/tvb/interfaces/web/controllers/flow_controller.py (hunk 7 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/controllers/flow_controller.py:258]: Done.

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html (hunk 9 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:11]: Done.

File: framework_tvb/tvb/core/entities/storage/datatype_dao.py [hunk 10]
```diff
@@ -518,6 +518,10 @@ def get_values_of_datatype(self, project_id, datatype_class, filters=None, page_
                         ).filter(DataType.invalid == False
                         ).filter(or_(Operation.fk_launched_in == project_id,
                                      Links.fk_to_project == project_id))
+
+            if hasattr(datatype_class, 'time_series_type'):
```
Comment (liadomide) [inline framework_tvb/tvb/core/entities/storage/datatype_dao.py:522]: This if is not sane here.
We should covert the cartesian product with BurstConfig in a left/right outer join on this condition

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html (hunk 9 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:9]: You say `done` but still you changed from `<p>` into `<div>` and did not give a justification why this change is necessary.

File: framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py (hunk 1 above)
Comment (liadomide) [reply framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py:309]: You resolved this, but the undesired change seem to be still here...

File: framework_tvb/tvb/interfaces/web/templates/jinja2/base_template.html [hunk 11]
```diff
@@ -47,6 +47,7 @@
         <script type="text/javascript" src="/static/jquery/jquery-ui.js?{{ currentVersionJS }}"></script> <!--JQuery custom imports for the Range, Slider controller -->
         <script type="text/javascript" src="/static/jquery/jquery.json.js"></script>
         <script type="text/javascript" src="/static/js/genericTVB.js?{{ currentVersionJS }}"></script>
+        <script type="text/javascript" src="/static/js/filters.js?{{ currentVersionJS }}"></script>
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/templates/jinja2/base_template.html:50]: This import place might be too high.
Have you tried to import this script in `genericAdapterFormFields.html` and in `main_burst.html` ?

File: framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html (hunk 9 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/templates/jinja2/form_fields/datatype_select_field.html:9]: Oh yes, I changed two <p>'s actually. The first one (the one from above in the code) can stay <p> but the second one not really... Because it won't include the <div>'s for the hidden form filters inside the second p, but outside them. I checked this on stackoverflow and in the html documentation it says that these kind of things can happen because inline elements are not supposed to contain block elements.

If we really wan't to keep it <p> then I need to find a different way to include the hidden filters new to the user defined ones.
image

File: framework_tvb/tvb/interfaces/web/templates/jinja2/base_template.html (hunk 11 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/interfaces/web/templates/jinja2/base_template.html:50]: If I put the inclusion in these two + spatial/includedResources.html (for the two stimulus types) then it's allright.

File: framework_tvb/tvb/core/entities/storage/datatype_dao.py (hunk 10 above)
Comment (VinczeRobert) [reply framework_tvb/tvb/core/entities/storage/datatype_dao.py:522]: Oh yes, because I changed it in two places actually. The one which is more above in th html file can stay <p> but the second one should be a div, otherwise I the hidden form filters will have to be handled in a different way than the user defined ones (I attached a picture to show what happens when I leave it as a p, basically the way the code is now, when you apply the filters, the effect of the form filters will be ignored). The HTML documentation says that such things (i.e. elements not appearing where you'd expect them) can happen because inline elements are not supposed to contain block elements. 

In conclusion, I either make it a div, or I have to make some changes to the way form filters are handled.

image

File: framework_tvb/tvb/core/entities/storage/datatype_dao.py [hunk 12]
```diff
@@ -512,12 +512,15 @@ def get_values_of_datatype(self, project_id, datatype_class, filters=None, page_
                                        func.max(text('"OPERATION_GROUPS_1".name')),
                                        func.max(DataType.user_tag_1)
                         ).join((Operation, datatype_class.fk_from_operation == Operation.id)
+                        ).outerjoin((BurstConfiguration, datatype_class.fk_from_operation ==
+                                     BurstConfiguration.fk_simulation)
```
Comment (liadomide) [inline framework_tvb/tvb/core/entities/storage/datatype_dao.py:516]: How about: 
`.outerjoin(BurstConfiguration, DataType.fk_parent_burst == BurstConfiguration.id)`
?

File: framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py [hunk 13]
```diff
@@ -53,6 +53,7 @@
 from tvb.config.init.introspector_registry import IntrospectionRegistry
 from tvb.interfaces.web.controllers.autologging import traced
 from tvb.interfaces.web.controllers.burst.base_controller import BurstBaseController
+from tvb.interfaces.web.controllers.common import KEY_ADAPTER
```
Comment (liadomide) [inline framework_tvb/tvb/interfaces/web/controllers/simulator_controller.py:56]: This import must be orphan, thus it can be removed right ?
